v1.0.0-rc.1: High level invocation spec (#21)
ucan-wg/invocation: README.md

# UCAN Invocation Specification v1.0.0-rc. 1

FIXME
- move pipelines, await, etc to own spec

> - explain in FAQ why we don't need `join` (i.e. use promises, since the model is inverted)

Reviewer: Ah, nice! Could `fork` be expressed with promises as well?
Wondered about this idea of expressing effects via awaits [down below](https://github.com/ucan-wg/invocation/pull/21/commits/ee652ef705c2050820fc4405e894499d0ed83732#r1345687658).

Author: No, this is communication back to the Invoker, so it's not signed. We can subsume `join` by what used to be called `fork`, though.

Reviewer: > Could fork be expressed with promises as well?
>> No, this is communication back to the Invoker, so it's not signed.

Aha, I see, thanks for clarifying!

> Enqueued [Task](https://github.com/ucan-wg/invocation/tree/high-level#32-task)s describe requests for future work to be performed.

These Tasks seem to be an interpretation of the Result.
E.g., given Result for a query returned 100 solutions => send an email.
I.e., enqued Tasks are some further actions based on / derived from the Result.

This interpretation / `deriveFurtherActions(Result)` could happen as part of the original Task, and it seems possible to perform it outside of it as well. Would it be valuable to have it outside, decoupling a Task from interpretation of its Result?

> - add expiry

Reviewer: As "this task is valid for a certain period of time"?
If so, would it be handy to express it at the UCAN level?

I.e., I want this Task invoked and UCAN execute it in that time frame

Author: I used to think so too, but we need separate expirations in this version. Because we more rigorously "close the loop", we're not delegating to the executor — the executor is the root issuer. This means that we can no longer infer the expiry from the enclosed delegation.

Reviewer: Hmm, I see how delegation to the root issuer doesn't grant him any new capabilities. Yet such delegation may still be useful to express "you can perform that specific attenuated stuff **on my behalf** in **that timeframe**"? An [example usage](https://github.com/ucan-wg/spec/issues/130#issuecomment-1762863709) (`exp` is omitted in those UCANs, but we can imagine that Invoker attenuated to a narrower time range). And so a valid Receipt would need to include explicit delegation from Invoker. :thinking:

Reviewer: Happy to learn more about why delegation to the root issuer is not a good approach, btw!

It may not be a good solution for specifying validity timeframe of invocation, if delegation UCAN with that validity timeframe is decoupled from the Invocation / comes out-of-band, as in the example above. Because then even when UCAN expires the Invocation hangs around. User then may issue the same Invocation, and its UCAN can be accedintally used for the previous Invocation as well.

```
at t1, UCAN1 for [t1; t3) -can-> Task <-run- Invocation1
at t3, UCAN2 for [t3; t6) -can-> Task <-run- Invocation2
```
say
at t3, Invocation still wasn't executed, and so it's meant to be considered expired, however
at t3, Issuer issued another Invocation for the same Task
and so
throughout [t3; t6) Executor now can use UCAN2 as authorization to execute both Invocation2 and _Invocation1_!
...Ah, they're actually the same, content-wise, Invocation1 = Invocation2.

Buuut if they would differ, say contained "decision time" - when they've been issued, akin to `iat` that's on Receipt - then Tasks would be the same, yet Invocations would be unique.
```
at t1, UCAN1 for [t1; t3) -can-> Task <-run- Invocation1 iat t1
at t3, UCAN2 for [t3; t6) -can-> Task <-run- Invocation2 iat t3
```

And so Executor could execute both Invocations, whereas Invocation1 is meant to be expired.

Perhaps an interesting solution to that is to have validity timeframe inside the Task.
And so both UCAN and Invocation would be talking about the same stuff.
```
at t1, UCAN1 -can-> Task for [t1; t3) <-run- Invocation1 iat t1
at t3, UCAN2 -can-> Task for [t3; t6) <-run- Invocation2 iat t3
```
and so
at t3, both UCAN1 and Invocation1 are expired
at t5, only UCAN2 and Invocation2 are valid

I.e., move valid timeframe from U of UCAN to CAN (Capability) it's about, so it can be used across both UCAN and Invocation, CAN()
I.e., validity timeframe as attenuation of Capability

> - Consider ability heirarcy as being subpaths: `fs/mutate/write`

Reviewer: Issuing a UCAN with e.g., `fs` would be kinda akin to issuing UCAN that bundles multiple single capabilities (`fs` `mutate` write`).
From what I heard, such UCANs are harder to manage due to lack of fine-granularity.
[It's been reported](https://github.com/ucan-wg/spec/issues/171#issuecomment-1733959444) that having single capability per UCAN is a handy approach. Just food for thought

Author: It's true, but also assuems that all agents are live processes. Adding new messages to a heirarchy ina bardwards-compatible way after the fact in an API is standard practice. Without this mechanism, you would need to issue new UCANs to every existing delegation chain.

Fission and DAG House both have use cases where the original Agent burned their keys after issuing, or is kept in cold storage. Reissuing is impossible in these cases.

Akiko is going away for the weekend. Her good friend Boris is going to borrow her car while she's away. They meet at a nearby cafe, and Akiko hands Boris her car keys. Boris now has the capability to drive Akiko's car whenever he wants to. Depending on their plans for the rest of the day, Akiko may find Boris quite rude if he immediately leaves the cafe to go for a drive. On the other hand, if Akiko asks Boris to run some last minute pre-vacation errands for that require a car, she may expect Boris to immediately drive off.

To put this in terms closer to a UCAN flow:

``` mermaid
sequenceDiagram
    participant 🚗
    actor Akiko
    actor Boris

    autonumber

    Note over 🚗, Akiko: Akiko buys a car
    🚗 -->> Akiko: Delegate(Drive 🚗)

    Note over Akiko, Boris: Boris offers to run errands for Akiko
    Boris -->> Akiko: Delegate(Boris to run errands)

    Note over Akiko, Boris: Akiko gives Boris access to her car
    Akiko -->> Boris: Delegate(Drive 🚗)

    Note over 🚗, Boris: Akiko asks Boris to use her car to run errands
    Akiko ->> Boris: Invoke!(Boris to run errands, using 🚗 (➌))

>     Boris ->> 🚗: Invoke!(Drive 🚗)

Reviewer: ```suggestion
    Boris ->> 🚗: Execute!(➍)
```

```

In the example above, steps ➌ and ➍ are qualitatively different:

- Step ➌ grants authority (to drive the car)

> - Step ➍ is a _command_ to do so

Reviewer: ```suggestion
- Step ➍ is a _request_ to do so
```
To not have name collision with `Command` of this spec

Reviewer: Or maybe
- Step ➌ authorizes capability (to drive the car)
- Step ➍ requests to _excercise_ it

| [Metadata]  | `meta`  | `{String : Any}`   | No          | Arbitrary [Metadata]                                               |
| [Nonce]     | `nonce` | `Bytes`            | No          | A unique, random nonce                                             |
| [Expiry]    | `exp`   | `Integer`[^js-num] | No          | The timestamp at which the Invocation becomes invalid              |
| [Issued At] | `iat`   | `Integer`[^js-num] | No          | The timestamp at which the Invocation was created                  |
| [Cause]     | `cause` | `&Receipt`         | No          | An OPTIONAL CID of the [Receipt] that enqueued the [Task]          |

The shape of the `args` MUST be defined by the `do` field type. This is similar to how a method or message contain certain data shapes in object oriented or actor model languages respectively. Using the JavaScript analogy from the introduction, an Action is similar to wrapping a call in a closure:

```js
// Command
{
  "do": "msg/send",
  "args": {
    "from": "mailto:alice@example.com",
    "to": [ "bob@example.com", "carol@example.com" ],

>     "subject": "hello",

Reviewer 2: Discussed on today's call-- someone CTRL-Fing for `subject` could get tripped up, so a quick rename to `emailSubject` or `messageTitle` or same might be a quick fix.

| `uci` | `&InvocationPayload` | Yes      | The CID of the [Invocation Payload]                      |
| `sig` | `&Signature`         | Yes      | A signature by the Payload's `iss` over the CID in `uci` |

## 3.4 Examples

### 3.4.1 Interacting with an HTTP API

```js
{
  "sig": {"/": {bytes: "7aEDQIscUKVuAIB2Yj6jdX5ru9OcnQLxLutvHPjeMD3pbtHIoErFpo7OoC79Oe2ShgQMLbo2e6dvHh9scqHKEOmieA0"}},
  "inv": {                                                                    //           ┐
    "iss": "did:plc:ewvi7nxzyoun6zhxrhs64oiz",                                //           │
    "aud": "did:key:z6MkrZ1r5XBFZjBU34qyD8fueMbMRkKw17BZaq2ivKFjnz2z",        //           │
    "run": cid({                                                              //  ┐        │
      "act": cid({                                 // ┐                           │        │
        "nnc": "&NCC-1701-D*",                     // │                           │        │
        "cmd": "crud/create",                      // │                           │        │
        "arg": {                                   // │                           │        │

>           "uri": "https://example.com/blog/posts", // │                           │        │

Reviewer 3: obviously, these args are different for different invocations. Can we standardize on a URI field that is Option<Uri> for any arg. It seems important enough that it can't just be treated as an any? I could be wrong, but it seems to be important generally.

### 3.4.2 Sending Email

```js
{
  "sig": {"/": {bytes: "7aEDQIscUKVuAIB2Yj6jdX5ru9OcnQLxLutvHPjeMD3pbtHIoErFpo7OoC79Oe2ShgQMLbo2e6dvHh9scqHKEOmieA0"}},
  "inv": cid({
    "iss": "did:plc:ewvi7nxzyoun6zhxrhs64oiz",
    "aud": "did:key:z6MkrZ1r5XBFZjBU34qyD8fueMbMRkKw17BZaq2ivKFjnz2z",
    "run": cid({
      "act": {
        "nnc": "email-akiko#1234567890"
        "cmd": "msg/send",
        "arg": {
          "from": "mailto:akiko@example.com",
          "to": [ "boris@example.com", "carol@example.com" ],
          "subject": "Coffee",
          "body": "Let get coffee sometime and talk about UCAN Invocations!"
        }
      }),
      "mta": {},
      "prf": [{"/": "bafkr4iblvgvkmqt46imsmwqkjs7p6wmpswak2p5hlpagl2htiox272xyy4"}],
      "exp": 1697409438
    }
  })
}
```

### 3.4.3 Inline WebAssembly

```js
{
  "sig": {"/": {bytes: "7aEDQIscUKVuAIB2Yj6jdX5ru9OcnQLxLutvHPjeMD3pbtHIoErFpo7OoC79Oe2ShgQMLbo2e6dvHh9scqHKEOmieA0"}},
  "inv": cid({
    "iss": "did:plc:ewvi7nxzyoun6zhxrhs64oiz",
    "aud": "did:key:z6MkrZ1r5XBFZjBU34qyD8fueMbMRkKw17BZaq2ivKFjnz2z",
    "run": ({
      "act": {
        "nnc": "", // NOTE: as stated above, idempotent Actions should always have the same nonce

>         "act": "wasm/run",

Reviewer 3: act within act is weird IMO. Shouldn't this be the `cmd` and that is something we can match against with any other invocation? It feels like something that should be standard so we don't have to be like look inside act, is it a cmd, or another act, etc.

[Public Resources]: #public-resources

A core part of UCAN's design is interacting with the wider, non-UCAN world. Many resources are open to anyone to access, such as unauthenticated web endpoints. Unlike UCAN-controlled resources, an invocation on public resources is both possible, and a hard requirement for initiating a flow (e.g. sign up). These cases typically involve a reference passed out of band (such as a web link). Due to [designation without authorization], knowing the URI of a public resource is often sufficient for interacting with it. In these cases, the Executor MAY accept Invocations without having a "closed-loop" proof chain, but this SHOULD NOT be the default behavior.

## Promise Pipelining
[Promise Pipelining]: #promise-pipelining

[UCAN Promise] extends UCAN Invocation with [distributed promise pipelines]. Promises are helpful in a wide variety of situations for efficiency and convenience. Implementations supporting UCAN Promises is RECOMMENDED.

# Concepts
[Concepts]: #concepts

## Roles
[Roles]: #roles

Task adds two new roles to UCAN: invoker and executor. The existing UCAN delegator and delegate principals MUST persist to the invocation.

> | UCAN Field | Delegation                             | Invocation                      |

Reviewer 2: > The existing UCAN delegator and delegate principals MUST persist to the invocation.

I'd love this to be unpacked a bit, it's too terse for me.  Are you saying that the _same identifiers_ used from principles at delegation time must be _dereferenceable_ at invocation time? I feel like phrasing that in terms of local context or dereferencing is prefereable to making it sound more existential (and untestable) as some kind of... survival in a hostile [distributed] environment 😅

    Invocation ||--|| Receipt: returns
    Receipt |o--|{ Task: enqueues
```

## Anatomy
[Anatomy]: #anatomy

| Concept      | Description                                                                 |
|--------------|-----------------------------------------------------------------------------|
| [Command]    | Function application; a description of work to be performed                 |
| [Task]       | Contextual information for a [Command], such as resource limits             |
| [Invocation] | A request to perform some [Task] based on [delegated][Delegation] authority |

A request for some work to be done (or to "exercise your authority") is an Invocation.

``` mermaid
flowchart TD
    subgraph Invocation
        SignatureBytes["Signature (raw bytes)"]
      
        subgraph SigPayload ["Signature Payload"]
            VarsigHeader["Varsig Header"]

            subgraph InvocationPayload ["Invocation Payload"]
                iss
                sub
                do
                args
                prf
                cause["cause (optional)"]
                etc["..."]
            end
        end
    end
    
    cause -.->|CID| Receipt
```

> As [noted in the introduction][lazy-vs-eager], there is a difference between a reference to a function and calling that function. The [Invocation] is a request to the [Executor] to perform the enclosed [Task]. [Invocation Payload]s are not executable until they have been signed and [Delegation] proofs validated.

Reviewer 2: ```suggestion
As [noted in the introduction][#lazy-vs-eager-evaluation], there is a difference between a reference to a function and calling that function. The [Invocation] is a request to the [Executor] to perform the enclosed [Task]. [Invocation Payload]s are not executable until they have been signed and [Delegation] proofs validated.
```

Note that the Invocation MUST include the Signature envelope. An [Invocation Payload] on its own MUST NOT be considered a valid Invocation.

> # [UCAN Envelope] Configuration

Reviewer 2: ```suggestion
# [UCAN Envelope Configuration]
```

[UCAN Envelope Configuration]: #ucan-envelope-configuration

## Type Tag
[Type Tag]: #type-tag
 

> The UCAN envelope's payload tag MUST be `ucan/inv@1.0.0-rc.1`.

Reviewer 2: ```suggestion
The UCAN envelope's [payload tag](https://github.com/ucan-wg/spec/blob/main/README.md#envelope) MUST be `ucan/inv@1.0.0-rc.1`.
```
Note that this^ URL won't be valid until [spec#173](https://github.com/ucan-wg/spec/pull/173) is merged, which includes the envelope version string.  (When do be bikeshed that field name, btw, because even writing this comment required a LOT of thought).

Author: Thanks! Merged manually to keep external links at the bottom of the doc

| `iss`   | `DID`              | Yes      | The DID of the [Invoker]                                           |
| `sub`   | `DID`              | Yes      | The [Subject] being invoked                                        |
| `aud`   | `DID`              | No       | The DID of the intended [Executor] if different from the [Subject] |
| `cmd`   | `String`           | Yes      | The [Command]                                                      |
| `args`  | `{String : Any}`   | Yes      | The [Command]'s [Arguments]                                        |
| `prf`   | `[&Delegation]`    | Yes      | [Delegation]s that prove the chain of authority                    |
| `meta`  | `{String : Any}`   | No       | Arbitrary [Metadata]                                               |
| `nonce` | `Bytes`            | No       | A unique, random nonce                                             |
| `exp`   | `Integer`[^js-num] | No       | The timestamp at which the Invocation becomes invalid              |
| `iat`   | `Integer`[^js-num] | No       | The timestamp at which the Invocation was created                  |

> | `cause` | `&Receipt`         | No       | An OPTIONAL CID of the [Receipt] that enqueued the [Task]          |

Reviewer 2: [Receipt] is also not defined anywhere... self-explanatory perhaps, but a place-holder section with a one-sentence summary would be good for now?

 
[^js-num]: JavaScript has a single numeric type ([`Number`][JS Number]) for both integers and floats. This representation is defined as a [IEEE-754] double-precision floating point number, which has a 53-bit significand.

The shape of the `args` MUST be defined by the `cmd` field type. This is similar to how a method or message contain certain data shapes in object oriented or actor model languages respectively. Using the JavaScript analogy from the introduction, an Action is similar to wrapping a call in a closure:

```js
// Command
{
  "cmd": "/msg/send",
  "args": {
    "from": "mailto:alice@example.com",
    "to": [ "bob@example.com", "carol@example.com" ],
    "subject": "hello",
    "body": "world"
  }
}
```

```js
// Pseudocode JS Analogy
() => msg.send({
  from: "mailto:alice@example.com",
  to: ["bob@example.com", "carol@example.com"],
  subject: "hello",
  body: "world"
})
```

### Agents
[Agents]: #agents

#### Issuer
[Issuer]: #issuer

> The `iss` field MUST include the Issuer of the Invocation. This DID MUST match against the encoding signature.

Reviewer 2: ```suggestion
The `iss` field MUST include the Issuer of the Invocation. This DID URL MUST dereference to the public key which proofs the signature over the payload included in the envelope.
```
As I mentioned on the top-level spec, a `did:key` doesn't need a fragment identifier because it always references to exactly one public key, but this is rare, most DID methods require you to express a qualified DID URL (i.e., a DID plus `#key1` or `#assertionMethod`, etc.) to get a specific key out of the DID Document.

Author: Manually ported so that I wouldn't forget to fix the typo in the `suggestion`. Thanks!

#### Subject
[Subject]: #subject

The REQUIRED `sub` field both parameterizes over a specific agent, and acts as a namespace for how to interpret the [Command]. This is especially critical for two parts of the life cycle:

1. Specifying a particular `sub` (and thus `aud`) when [enqueuing new Tasks][enqueue] in a Receipt
2. Indexing Receipts for reverse lookup and memoization

#### Audience
[Audience]: #audience

> The OPTIONAL `aud` field specified the intended recipient of Invocation, otherwise the Audience MUST be assumed to the [Subject]. This is useful for message routing, command brokers, proxy execution, gateways, replicated state machines, and so on.

Reviewer 2: But `aud` is still [REQUIRED in the high-level UCAN spec](https://github.com/ucan-wg/spec/blob/9b50374ba600eea56a697f204f870cd36fba37c7/README.md#payload)?

### Task
[Task]: #task

A Task is the subset of Invocation fields that uniquely determine the work to be performed. The nonce is important for distinguishing between non-idempotent executions of a Task by making the group together unique.
 

> A Task MUST be uniquely defined by a Task ID that is the CID of the following fields:

Reviewer 2: of the... CONCATENATION of these fields? of a flat structure of all their values? I think being a little more explicit in how these fields get canonicalized before get magically CID-ified would go a long way here.

Author: It's the CID of those fields as a struct. Will fix.

- [Subject]
- [Command]
- [Arguments]
- [Nonce]

Tasks that describe pure functions — or other strategies like fan-out racing — SHOULD have the same Task ID by using the same nonce.

#### Command
[Command]: #command

The REQUIRED Command (`cmd`) field MUST contain a concrete, dispatchable message that can be sent to the Executor. The Command MUST define the shape of the data in the [Arguments].

### Arguments
[Arguments]: #arguments

The REQUIRED Arguments (`args`) field, MAY contain any parameters expected by the Command. The Subject MUST be considered the authority on the shape of this data. This field MUST be representable as a map or keyword list.

> The Arguments MUST pass validation of the Policies on all of the [UCAN Delegations][UCAN Delegation] in the [Proofs] field. If any Policy reports failure, the Invocation MUST be rejected.

Reviewer 2: Note that [Arguments] is not defined in the high-level spec either, and there are still references to "Policy" there worth CTRL-Fing to clean up once Arguments is added there (as optional, I assume?)

Author: Arguments is not a common field across all UCANs. It's special for Invocation.

#### Nonce
[Nonce]: #nonce

The REQUIRED `nonce` field MUST include a random nonce. This field ensures that multiple (non-idempotent) invocations are unique. The nonce SHOULD be empty (`0x`) for Commands that are idempotent (such as deterministic Wasm modules or standards-abiding HTTP PUT requests).

### Proofs
[Proofs]: #proofs

The `prf` field defines all [Delegation]s required to prove that this Invocation has an unbroken authorization chain.

### Provenance
[Provenance]: #provenance

#### Cause
[Cause]: #cause

The OPTIONAL `cause` field is a provenance claim describing which [Receipt] requested it. This is helpful for tracking chains of Invocations.

#### Expiration
[Expiration]: #expiration

> The OPTIONAL field `exp` defines when the Invocation SHOULD time out. This is both expressive (defines a timeout, which is a best practice), and prevents replays.

Reviewer 2: But `exp` is still [REQUIRED in the high-level UCAN spec](https://github.com/ucan-wg/spec/blob/9b50374ba600eea56a697f204f870cd36fba37c7/README.md#payload)?

Author: Fixed

#### Issued At
[Issued At]: #issued-at

The OPTIONAL `iat` field MAY contain an issuance timestamp. This time SHOULD NOT be trusted; it is only a claim by the Invoker of their system time. System clocks often have clock skew, or a Byzantine Invoker could claim an arbitrary time.

#### Metadata
[Metadata]: #metadata

> The OPTIONAL `meta` field MAY include arbitrary metadata or extensible fields. For example, Wasm fuel, an internal job ID, references to GitHub Issues, and so on. This data MAY be used by the Executor.

Reviewer 2: this "MAY be used" is a little confusing relative to this passage in the [high level spec](https://github.com/ucan-wg/spec/blob/9b50374ba600eea56a697f204f870cd36fba37c7/README.md#metadata): 

>  The data contained in this map MUST NOT be semantically meaningful to delegation chains.

Author: This isn't for delegation chains

[Marc-Antoine Parent]: https://github.com/maparent
[Mark Miller]: https://github.com/erights
[OAuth 1]: https://oauth.net/1/
[OCapN]: https://github.com/ocapn/
[Oxide Computer]: https://oxide.computer/ 
[Philipp Krüger]: https://github.com/matheus23/
[Protocol Labs]: https://protocol.ai
[Quinn Wilton]: https://github.com/QuinnWilton
[Robust Composition]: http://www.erights.org/talks/thesis/markm-thesis.pdf
[Rod Vagg]: https://github.com/rvagg/
[Simon Worthington]: https://github.com/simonwo
[Spritely Institute]: https://spritely.institute/news/introducing-a-distributed-debugger-for-goblins-with-time-travel.html
[UCAN Delegation]: https://github.com/ucan-wg/delegation/
[UCAN Promise]: https://github.com/ucan-wg/promise/
[URI]: https://en.wikipedia.org/wiki/Uniform_Resource_Identifier
[Witchcraft Software]: https://github.com/expede
[Zeeshan Lakhani]: https://github.com/zeeshanlakhani
[`data`]: https://en.wikipedia.org/wiki/Data_URI_scheme
[`ipfs`]: https://docs.ipfs.tech/how-to/address-ipfs-on-web/#native-urls
[`magnet`]: https://en.wikipedia.org/wiki/Magnet_URI_scheme
[capability systems]: https://en.wikipedia.org/wiki/Capability-based_security

> [designation without authorization]: https://srl.cs.jhu.edu/pubs/SRL2003-02.pdf

Reviewer 2: ```suggestion
[designation with authorization]: https://srl.cs.jhu.edu/pubs/SRL2003-02.pdf
```
This is hard to grok, since in that article there is a principle of "_NO_ designation without authority", inverted when applied as [an e2e design principle] to become "designation with authority" on page 3.

``` mermaid
flowchart RL
    subgraph Invocation
        direction LR
        subgraph InvocationPayload
            subgraph Task
                direction TB

                Command
                DelegationProofs[Delegation Proofs]
            end
        end

        Inv1Sig[Signature]
    end

    subgraph Receipt
        subgraph ReceiptPayload
            Ran
            Result
            Enqueue
        end

        RecSig[Signature]
    end

    subgraph Invocation2["(Next) Invocation"]
        direction RL

        subgraph InvocationPayload2[Invocation Payload]
            subgraph Task2
                Command2["Command"]
                DelegationProofs2[Delegation Proofs]
            end

            Cause
        end

        Inv2Sig[Signature]
    end

    Cause -->|provenance| Enqueue
    Ran   -->|provenance| Invocation
```

# 3 Request

## 3.1 Action

A Command is the smallest unit of work that can be Invoked. It is akin to a function call or actor message. It MUST conform to the following struct shape:

| Name        | Field | Type             | Required |
|-------------|-------|------------------|----------|

> | [Subject]   | `sub` | `DID`            | No       |

Author: Ah, perhaps this is confusing. You need to align the `delegation.sub` with the `invoctaion.aud`. I see why that may not be ideal.

> | [Command]   | `cmd` | `String`         | Yes      |

Author: Hmm yeah now that we have a single resource per invocation, this is a thing we could do 🤔 I guess it doesn't work if your proof chain starts with a `ucan/*` or equivalent (to connect disparate chains). Will think on it, but really we're talking about keeping or eliminating some redundancy.

> | [Command]   | `cmd` | `String`         | Yes      |

Author: Hmm you know, `sub` now imposes a `context` (in the W3C/RDF sense) for the `cmd` and `args`... mayeb it should always be required as the authority. So things like Wasm (which can be run by anyone) would just have a standard `sub` set by the UCAN WG.

  to: ["bob@example.com", "carol@example.com"],
  subject: "hello",
  body: "world"
})
```

### 3.1.1 Command

The Command (`cmd`) field MUST contain a concrete, dispatchable message that can be sent to the Executor. The Command MUST define the shape of the data in the [Arguments].

This field can be thought of as the method, message, or trait being sent to the resource. Note that _unlike_ a UCAN Delegation [Ability], which includes hierarchy, a Command MUST be fully concrete.

### 3.1.2 Arguments

The Arguments (`arg`) field, MAY contain any parameters expected by the Command. The Subject MUST be considered the authority on the shape of this data. This field MUST be representable as a map or keyword list.

UCAN capabilities provided in proofs MAY impose certain constraint on the type of Arguments allowed.

### 3.1.3 Subject

The OPTIONAL `sub` field is intended for cases where parameterizing a specific agent is important. This is especially critical for two parts of the life cycle:

1. Specifying a particular `sub` (and thus `aud`) when [enqueuing new Tasks][enqueue] in a Receipt
2. Indexing Receipts for reverse lookup and memoization

### 3.1.4 Nonce

> If present, the REQUIRED `nnc` field MUST include a random nonce expressed in ASCII. This field ensures that multiple (non-idempotent) invocations are unique. The nonce SHOULD be `""` for Commands that are idempotent (such as deterministic Wasm modules or standards-abiding HTTP PUT requests).

Author: Ooh this is just totally out of date. Will fix.

## 3.2 Task

A Task wraps a [Command] with contextual information. This includes expiration time, delegation chain, and extensible metadata for things like resource limits.

| Field | Type                       | Required | Description                                                                    |
|-------|----------------------------|----------|--------------------------------------------------------------------------------|

> | `act` | `&Command`                 | Yes      | The CID of the [Task] to be run                                                |

Author: Fixed

| Field | Type              | Required | Description                                                              |
|-------|-------------------|----------|--------------------------------------------------------------------------|
| `ucr` | `&ReceiptPayload` | Yes      | The CID for the [Receipt Payload]                                        |
| `sig` | `Signature`       | Yes      | The [Signature] of the `uci` value, by the [Receipt Payload]'s `iss` DID |

## 4.2 Receipt Payload

Receipt Payloads MUST conform to the following shape:

| Field | Type               | Required | Description                                                                        |
|-------|--------------------|----------|------------------------------------------------------------------------------------|
| `ucr` | `SemVer`           | Yes      | The version of this spec that the Receipt conforms to                              |
| `iss` | `DID`              | Yes      | The DID of the Executor                                                            |
| `ran` | `&Invocation`      | Yes      | A link to the [Invocation] that the Receipt is for                                 |

> | `prf` | `[&Delegation]`    | Yes      | [Delegation] proof chain if the Executor was not the `aud` of the `ran` Invocation |

Author: Interesting! Let me mull that over

| `out` | `Result`           | Yes      | The value output of the invocation in [Result] format                              |
| `enq` | `[&Task]`          | Yes      | Further [Task]s that the [Invocation] would like to enqueue                        |
| `mta` | `{String : Any}`   | Yes      | Additional data about the receipt                                                  |

> | `rec` | `&Receipt`         | No       | Recursive `Receipt`s if the Invocation was proxied to another Executor             |

Author: 💡

## 2.2 Components

### 2.2.1 Instruction

An [Instruction] is like a deferred function application: a request to perform some action on a resource with specific input.

### 2.2.2 Task

A [Task] wraps an [Instruction] with runtime configuration, including timeouts, fuel, trace metadata, and so on.

### 2.2.2 Authorization

> An [Authorization] is a cryptographically signed proof permitting execution of referenced [Task]s. It allows the [Invoker] to authorize a group of tasks using one cryptographic signature.

Author: Yep I know. It's a `draft`.

### 2.2.3 Invocation

> An [Invocation] is an [Authorized] [Task].

Author: Yes, intentionally. It doesn't make sense to get these to all line up so early in the Draft PR process, because you end up spending a ton of time making links work as stuff moves around the doc.

## 3.1 Schema

```
type Task struct {
  rsc   URI
  op    Ability
  input {String : Any}

>   nnc   String 

Reviewer: Nonce at the level of Instruction makes them less reusable in `await`s.
They do make Instructions a more unique, and may be thought to be used in order for an Instruction to await some particular Receipt of other Instruction.
E.g., in Effects that depend on Result of the Receipt that caused them.
However, multiple Reciepts for the same Nonced Instruction still may exist, so it can't be used to uniquely identify a particular Receipt.

Perhaps the problem of "resolve await to this particular Receipt" could be addressed at the level of Task.
[Effects could be Instructions](https://github.com/ucan-wg/invocation/pull/21/files#r1316855451), and Tasks created for them could capture exact Receipt "awaited".
I.e., Tasks would become ala closures.
It's been previously proposed to capture "awaited" at the level of Recipes #22, some reasons for that listed there, perhaps doing at Task/Closure level is more ellegant.
```json
{
 "bafy...someInstruction":
 {"input": {"someAwaitedVal": {"await/ok": {"/": "bafy...instructionToAwait"}}}},

 "bafy...someInstructionTask":
 {"run": {"/": "bafy...someInstruction"},
  "awaited": [{"/": "bafy...instructionToAwaitReceipt"}]}
}
 ```

Author: > Nonce at the level of Instruction makes them less reusable for awaits.

Unique invocations are a hard requirement if you want to be able to schedule tasks easily, reuse results, and do semantic awaits. For deterministic wasm, the nonce should be set to just `""`, but te general case of effectful tasks need to have unique IDs per call, since they don't catpure their entire context in the invocation.

```json
{
  "scope": [
    {"/": "bafyreihtmwju3okftpeuqe3x3ux5e7c2jescakwnoiyv45vnicke4kdxy4"},
    {"/": "bafyreieuo63r3y2nuycaq4b3q2xvco3nprlxiwzcfp4cuupgaywat3z6mq"}
  ],
  "s": {"/": {"bytes": "7aEDQIJB8XXJ6hWbwu40fN4bq8+Zq8BxyybSWXatMVU3VsL+yzVYpeJqsEBQE5rNtUJefR5rRCNimKNZMJjA9/udZQQ"}}
}
```

# 5 Task

As [noted in the introduction][lazy-vs-eager], there is a difference between a reference to a function and calling that function. The [Invocation] is an instruction to the [Executor] to perform enclosed [Task]. [Invocation]s are not executable until they have been provided provable authority (in form of UCANs in the `prf` field) and an [Authorization] (in the `auth` field) from the [Invoker].

The `auth` field MUST be contain an [Authorization] which signs over the `&Task` in `run`.

Concretely, this means that the `&Task` MUST be present in the associated `auth`'s `scope` field. A `Receipt` where the associated [Authorization] does not include the [Task] in the `scope` MUST be considered invalid.

## 5.1 Task

```
type Task struct {
  run     &Instruction
  meta    {String : Any}
  prf     [&UCAN]

  # Receipt of the invocation that caused this invocation

>   cause   optional &Receipt

Reviewer: Will this create a recursive link chain between the issued in fx Task and the Reciept it's been issued from?
```
Reciept -> Result -> Task
      ^              |
      ^--------------
      ```
Worrying, because such a chain seems to be impossible to express in IPLD

Reviewer: Perhaps one way to preserve causality yet have content-addresing is to introduce a separate Fx Reciept, from which joined Invocations can refer to Result Reciept
![image](https://github.com/ucan-wg/invocation/assets/13056889/8449cad4-8665-411f-ae54-8485d33a71c8)

Reviewer: Perhaps yet another way is to refer from Fx to Instruction, and capture causality in Invocation of that Instruction

![image](https://github.com/ucan-wg/invocation/assets/13056889/4077bd19-76ea-47b1-ae84-73dc76c85f52)

Reviewer: Another iteration of thoughts,
seems enough to:
1. have Instructions as Effects
2. capture causality in Tasks (as it is now)

![image](https://github.com/ucan-wg/invocation/assets/13056889/053849d5-6fd9-4642-bccc-5cf48a9e9bd7)

Let me know if it's bonkers :smile:

Reviewer: To find a solution we could start from defining the problem.
From my understanding, we need a way to capture causality of subsequent invocations
From that standpoint, it seems causality could be kept at Invocation level, and Instruction that outputs Effects does not actually care about how it's configured (Task) and invoked, Invoker's in control of that.
So, perhaps we can have:
1. `Effect`s to be Instructions
2. How such Instructions is to be configured may be specified by the Invoker (in Authorization of the caused Invocation?), in the form of UCAN that authorizes Executor to issue further Tasks (with configuration, e.g., limiting gas) on behalf of Invoker.
3. How it's to be invoked may be specified in the same way
4. Executor issues Tasks and their Invocation on behalf of Invoker
5. Causality's captured at the level of Tasks, in the same way it is now

![image](https://github.com/ucan-wg/invocation/assets/13056889/ed7f66e1-2990-43ed-8038-0899057356cc)
`Effect`s depend on some run-time values from Instruction, but these values been baked-in into them at run-time, when they've been created by the Instruction.

---

There are other ways to provide `Effect`s with Result (note: Result is a subset of run-time values, which's meh already), but they'll only add complexity. Feel free to skip this section.
But if you're curious how we could complicate lives of Invoker and Executioner :smile:, we could either:
1) requiring Invoker to create effects as a function of Invocation, producing `Effect`s with await/ok on Instruction and require Executor to resolve it to the exact Receipt.
2) by requiring to create an Effect as a function of Result, baking it into `Effect`s.

These two seems to be redundant hussle.

They'd look like:
1)
![image](https://github.com/ucan-wg/invocation/assets/13056889/2bf2e78c-6a8b-48d2-9e2c-17e9ae1ec32f)

2) Here there's more work for the Executor, as it needs to specify the exact Receipt to be awaited somehow (and the spec needs to be extended to require that), and then, subsequently, he (or his delegatee) needs to resolve that value.
Another con is that, likely, it'll be of value for Effect-producing logic to have at least Result value (ideally run-time data, but we're short on that at this stage), 
![image](https://github.com/ucan-wg/invocation/assets/13056889/7e9c3dd9-879c-4359-a13a-1fe064f32f8e)
Here, the spec needs to be extended to require Executor to resolve await/ok to the `cause` Receipt.

Author: It's a good thing to raise, but it's not possible to create a cycle because of hashing. You'd have to find a hash cycle in SHA2-256, which is impractical.

You can emit a new copy of the invocation to emulate loops via recursion, but each will need a unique nonce and thus unique hash. This enqueuing behaviour lets a runtime get between each run and check if it wants to keep running, suspend, or move to a different machine, all while maintaining totality in each run.

Reviewer: Correct that cycles are impractical to have, and are not intended.
My worry is that the spec atm implies a cycle.
```
Receipt -> Effect -> Invocation
  ^                   |
  ^---------cause-----
```
Since [Effects are Invocations](https://github.com/ucan-wg/invocation/pull/21/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L780) that capture in `cause` the Receipt that caused them.

The suggested approach is to have Effects as Instructions. Then we won't have the cycle and it seems more semantically aligned with what Invoker means to express, I imagine he'd like to issue further Instructions and not care to create and invoke a task from within the caused Instruction (also that would require to have it's private key baked into the caused Instruction, as he would need to sign Task and Invocation), instead he issues Instruction and Executor creates Task and Invokation on Invoker's behalf.

Side-note, currently there is misalignment in spec about where `cause` resides, the commented line above has it under Task type, whereas in [here](https://github.com/ucan-wg/invocation/pull/21/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R422) it is described to reside under Invocation. Personally, `cause` at the Invocation level seems semantical.

}
```

### 5.1.1 Instruction

The `run` field MUST contain a link to the [Task] to be run.

### 5.1.2 Metadata

The OPTIONAL `meta` field MAY be used to include human-readable descriptions, tags, execution hints, resource limits, and so on. If present, the `meta` field MUST contain a map with string keys. The contents of the map are left undefined to encourage extensible use.

If `meta` field is not present, it is implicitly a `unit` represented as an empty map.

### 5.1.3 Proofs

The `prf` field MUST contain links to any UCANs that provide the authority to perform this task. All of the outermost proofs MUST have `aud` field set to the [Executor]'s DID. All of the outmost proofs MUST have `iss` field set to the [Invoker]'s DID.

### 5.1.4 Optional Cause

[Task]s MAY be invoked as an effect caused by a prior [Invocation]. Such [Invocation]s SHOULD have a `cause` field set to the [Receipt] link of the [Invocation] that caused it. The linked [Receipt] MUST have an `Effect` (the `fx` field) containing invoked [Task] in the `run` field.

## 5.2 (Signed) Invocation

An [Invocation] is a signed [Task].

```
type Invocation struct {

>   task     Task

Reviewer: Ah, and we do need to be able to issue Tasks independently from Invocations, and in examples `task` is used as &Task.
I guess it's a typo here

  auth     &Authorization
}
```

### 5.2.1 Task

The [Task] containing the [Instruction] and any configuration.

### 5.2.2 Authorization

The `auth` field MUST contain a [Varsig] signing over an array of CIDs that includes the [Task]'s CID, signed by the issuer of the proofs.

## 5.3 Invocation Tag

An invocation capsule associates the [Invocation] with a versioned schema. This field is NOT REQUIRED. Wrapping an [Invocation] in an Invocation Tag is RECOMMENDED in contexts where the schema and version are not clear from context, such as when it is being stores or passed around without being nested in another structure that defines the version in its schema.

```

> type InvocationTag union {

Reviewer: Atm Tags hang loose, outside the DAG. E.g., Reciept references Invocation and not Invocation Tag, leaving the Tag out. I.e.,
```
Receipt ------> Invocation -> Task -> Instruction
Invocation Tag -^
```
Same with Reciept Tag.
Perhaps we could extend any spec that may contain Tagged types with such support, e.g.,
```
type Receipt struct {
  ran     &Invocation | &InvocationTag
}
```

That seems to comlicate the spec and add another layer of wrapping (and CID resolution) (resolve Receipt, *resolve InvocationTag*, resolve Invocation), as well as making Invocation's context-dependent / not self-contained / not having sufficient description on their own.

Perhaps another way is to extend Invocation itself with an optional `version` field?
Akin to how it's done for UCANs [with `ucv`](https://github.com/ucan-wg/spec#321-version)?

Author: Yeah, this is all getting moved out of this spec and over to the IPLD representation (which is its own spec now). "Capsule types" are a pattern in that ecosystem, but we don't want it to spread through every possible representation.

}
```

### 4.2.2 Enqueue

The result of an [Invocation] MAY include a request for further actions to be performed. This is a process of requesting that the invoker "enqueue" a new Task. This enables several things: a clean separation of pure return values from requesting impure tasks to be performed by the runtime, and gives the runtime the control to decide how (or if!) more work should be performed.

Enqueued [Task]s describe requests for future work to be performed. They SHOULD come with [Delegation]s, but MAY be a simple request back to the Invoker.

> All [Task]s in an [enqueue] array MUST be treated as concurrent, unless explicit data dependencies between them exist via [UCAN Promise]s.

Author: > For what it's worth join was analogy to a [thread.join](https://doc.rust-lang.org/stable/std/thread/struct.JoinHandle.html#method.join)

Right, but that's not how `join` works. You don't chain `join`s together to continue computation. The analogy seems like a stretch to me.

> but here we lost something that some of our work started to depend on.

We can add back in a special field if there's a clear use case, but can you explain why you can't use the 0-index of the array.

At least _structurally_, these are equivalent:

```haskell
data Fx = [Task]

data Fx' = Fx' {
  cont :: Maybe Task
  fork :: [Task]
}
```

Since this is structured more like actors or dataflow, there isn't really a "main" thread, right?

> All [Task]s in an [enqueue] array MUST be treated as concurrent, unless explicit data dependencies between them exist via [UCAN Promise]s.

Reviewer 4: > You don't chain joins together to continue computation.

That is precisely what we do in workflows that process data through multiple queues. We usually have `piece/offer` task that will have .join on `piece/accept` which in turn will have `join` on `aggregate/offer`  etc...

> can you explain why you can't use the 0-index of the array.

If it's an array with one element does it imply that it is a continuation or concurrent task ? Whatever's answer how do I express case where I have not both but the other from one you've answered.

In other words we wan to express tree different cases `{fork}`, `{join}` and `{fork, join}` with `[Task, ...Task[]]`  we only get `[Join]`  or `[Join, ...Fork[]]` and can not get `[...Fork[]]` unless we add another signal (which is what the current structure did)

----

I suspect for IPVM `join` is essentially inferred from `await/...`'s, which makes sense. However in our experience it is a lot easier to UCAN-ify legacy systems if you can simply express `join`'s without having to invert the control flow.

  # When issuer is different from executor this MUST hold a UCAN
  # delegation chain from executor to the issuer. This should be 
  # omitted when the executor is the issuer.
  prf     [&UCAN]
  
  sig     Varsig
}
```

### 8.1.1 Ran Invocation

The `ran` field MUST include a link to the [Invocation] that the Receipt is for.

### 8.1.2 Output

The `out` field MUST contain the value output of the invocation in [Result] format.

### 8.1.3 Effect

The OPTIONAL `fx` field, if present MUST be set to the caused [Effect]. The [Executor] SHOULD invoke contained [Task] to progress a workflow execution.

If `fx` does not contain OPTIONAL `join` field, it denotes completion of the current execution thread.

### 8.1.4 Metadata Fields

The OPTIONAL metadata field MAY be omitted or used to contain additional data about the receipt. This field MAY be used for tags, commentary, trace information, and so on.

### 8.1.5 Receipt Issuer

The OPTIONAL `iss` field, if present MUST contain the DID of the [Executor] delegate that signed it. If field is present, delegation from [Executor] MUST be included in the `prf` field.

If `iss` field is omitted, Receipt MUST be signed by the [Executor].

> ### 8.1.6 Proofs

Reviewer: Hasn't Executor been already authorized to perform excution of an Invocation via the Invocation's `auth` field?

Reviewer: Ah, I see, Receipt Proofs are for Executor's delegate to proof it's been authorized by Executor
